frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java

```diff
@@ -208,13 +209,74 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
   }
 
   @UIEffect
-  private void openDiscoverDialog() {
-    var action = ActionManager.getInstance().getAction(ACTION_DISCOVER);
-    var dataContext = DataManager.getInstance().getDataContext(EnhancedGraphTable.this);
-    var anActionEvent = AnActionEvent.createFromDataContext(ACTION_PLACE_EMPTY_TABLE, new Presentation(), dataContext);
-    action.actionPerformed(anActionEvent);
+  private void doAutomaticDiscover(Path macheteFilePath) {
+    var selectedRepository = getGitRepositorySelectionProvider().getSelectedGitRepository().getOrNull();
+    if (selectedRepository == null) {
+      LOG.error("Can't do automatic discover because of undefined selected repository");
+      return;
+    }
+    var mainDirPath = GitVfsUtils.getMainDirectoryPath(selectedRepository).toAbsolutePath();
+    var gitDirPath = GitVfsUtils.getGitDirectoryPath(selectedRepository).toAbsolutePath();
+
+    @SuppressWarnings({"nullness", "guieffect"}) // To allow runResult be uninitialized and Tuple accept @Nullable IBranchLayout
+    var task = new Task.Backgroundable(project,
+        getString("string.GitMachete.EnhancedGraphTable.automatic-discover.task-title")) {
+      private Try<Tuple3<IBranchLayoutWriter, @Nullable IBranchLayout, IGitMacheteRepositorySnapshot>> runResult;
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        runResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+            .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot())
+            .filterTry(repositorySnapshot -> repositorySnapshot.getRootBranches().size() > 0,
+                __ -> new EmptyRepositoryException())
+            .map(repositorySnapshot -> {
+              var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
+              var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
+              return Tuple(branchLayoutWriter, branchLayout, repositorySnapshot);
```

Comment:
I'd rather create a dedicated private static class for the result... tuples with 3+ fields of different types are confusing